### DISCUSSION TITLE
NormalizeActions Behaviour

### DISCUSSION THREAD

**Discussion by Vextasy** [cinar/indicator#168] [Q&A]
I notice that the comment for strategy.NormalizeActions mentions that it will “ensure the order follows a pattern of Hold, Buy, Hold, Sell”.

A side effect of maintaining that pattern is that the output from it is not superficially intuitive.  For example calling it with the sequences below (on the left) give rise to the following outputs (on the right): 

```
[Hold Sell Buy] ——> [Hold Hold Buy]
[Hold Buy Sell] ——> [Hold Buy Sell]
[Sell Buy Sell] ——> [Hold Buy Sell]
[Sell Sell Sell] ——> [Hold Hold Hold]
[Hold Hold Sell] ——> [Hold Hold Hold]
```

They don’t quite ensure the pattern of Hold, Buy, Hold, Sell but what is quite surprising is that any initial Sell action is lost.
This caught me out when looking at the annotations on a strategy chart which was displaying a set of actions that happened to start with a Sell action - the result was that there was no Sell annotation displayed on the chart.

This will happen in all charts that use action.ActionsToAnnotations for generating their annotations as it calls NormalizeActions.

NormalizeAction looks like this:

```go
func NormalizeActions(ac <-chan Action) <-chan Action {
	last := Sell

	return helper.Map(ac, func(a Action) Action {
		if a != Hold && a != last {
			last = a
			return a
		}

		return Hold
	})
}
```

My initial thought was that it should simply begin by initialising ‘last’ to Hold (just like DenormalizeAction does) but I wondered if there was a subtle reason for its behaviour?

**Comment by Vextasy** [comment]
I note that the test for NormalizeActions expects that outcome. But it does seem to give some confusing effects in the charts.

**Comment by cinar** [comment]
Hi, I'm traveling so sorry for the late response.

Yes, the NormalizeActions function ensures that an asset is first bought before it can be sold.

So let's look at the examples again:

```
[Hold Sell Buy] ——> [Hold Hold Buy]
```

The input suggests holding, selling, and buying. Hold moves as is, but as the asset is not yet being bought, NormalizeAction replaces the Sell action with Hold.

```
[Hold Buy Sell] ——> [Hold Buy Sell]
```

This is already correctly ordered, so the function makes no changes to it.

```
[Sell Buy Sell] ——> [Hold Buy Sell]
```

Same here, as the asset is not yet being bought, it can't be sold, so the function replaces the first Sell action with an Hold.

It helps removing the repeated Buy and Sell actions from the strategies when annotating the charts, also makes the actions aligned with how the outcome gets calculated.

However I do see your point. I believe the Outcome function is already keeping track of its Buy/Sell ordering, so perhaps a small change here can make the earlier Sell actions visible on the chart without impacting the Outcome calculation.

Once I am back, I can check to see if I can make this change.

**Comment by cinar** [comment]
I believe we resolved this, but please reopen if you need to.
